Match annexscientificname with scientificname

damianooldoni opened the pull request:
This draft PR aims to end the work about matching names (scientificnames) in `annexscientificname` table with names (scientificnames) in `scientificname` table and so filling a valid `scientificnameId`. This is done by the function `match_annexscientificname_to_scientificname()` in [file with same name](https://github.com/inbo/speciesbim/blob/annexscientificname/scripts/match_annexscientificname_to_scientificname.py).

The non trivial problem to solve was the match by `scientificname` + `authorship` even when `authorship` was `null`.
I assigned the draft status to this PR because I still have an issue with names appearing more than one time in `annexscientificname`. They are added to `scientificname` the first time and a `scientificnameId` is assigned correctly. But I get a `NoneType` when trying to match the second time ([L29-L33](https://github.com/inbo/speciesbim/blob/annexscientificname/scripts/match_annexscientificname_to_scientificname.py#L29-L33)).

If I debug and perform the same from console I get no errors. Strange. @niconoe: can you review and so hopefully solve this bug? I think a couple of extra eyes can do the work. Thanks.

damianooldoni committed:
Link annexscientificname to scientificname instead of taxonomy

See https://github.com/inbo/speciesbim/issues/36#issuecomment-689448745

damianooldoni committed:
Link names in annexes to scientificname table instead of taxonomy

See https://github.com/inbo/speciesbim/issues/36#issue-693009123

damianooldoni committed:
Merge branch 'master' into annexscientificname

damianooldoni committed:
Do not insert speciesannex names in scientificname table

As we use names in csv file https://github.com/inbo/speciesbim/blob/annexscientificname/data/raw/official_annexes.csv

damianooldoni committed:
Improve one entry in official annex

damianooldoni committed:
Separate populating annexscientificname from matching with scientificname

damianooldoni committed:
Add patch unique constraint to go further

It's a temporary patch in order to go further in development while waiting for a stable solution (see #43)

damianooldoni committed:
Start a new line

Previous use of print() had flush=True

damianooldoni committed:
Correct typo in name column scientificnameId of annexscientificname

damianooldoni committed:
Uncomment the new function match_annexscientificname_to_scientificname

damianooldoni committed:
Add summary of match in print() and add to log

damianooldoni committed:
Remove commented and unneeded code

damianooldoni committed:
Add unmatched_only status to message and add to log

damianooldoni committed:
Renumber all steps after match annexscientificname to scientificname

damianooldoni committed:
Add filename before calling function match_annexscientificname_to_scientificname

damianooldoni committed:
Uncomment step 5

damianooldoni committed:
Uncomment all next steps

damianooldoni committed:
Add authorship column and start populating it

See https://github.com/inbo/speciesbim/issues/36#issuecomment-704216557

damianooldoni committed:
Add authorship while populating annexscientificname table

See https://github.com/inbo/speciesbim/issues/36#issuecomment-704216557

damianooldoni committed:
Add constraint about uniqueness combi scientificname - authorhsip

damianooldoni committed:
Add authorhsip to annexscientificname table while creating it

damianooldoni committed:
Add authorship to dict to write in table annexscientificname

damianooldoni committed:
Remove wrong constraint on scientificName only

damianooldoni committed:
Add functionality for matching annexscientificname to scientificname

damianooldoni committed:
Number the steps from 0

In this way drop tables is step 0

damianooldoni committed:
Update annex demo based on structure with authorship

damianooldoni committed:
Set other names duplicates proof

damianooldoni committed:
Update demo

damianooldoni committed:
Add link to transform_db.py script for dropping tables

damianooldoni committed:
Add partial idnex and improve INSERT-CONFLICT-SELECT statement

damianooldoni committed:
Modify name fields containing original names in annex

See https://github.com/inbo/speciesbim/issues/36#issuecomment-704465296

damianooldoni committed:
Move function update_scientificname_id below

damianooldoni committed:
Add function to remove aux columns from annexscientificname

damianooldoni committed:
Set camelCase for printed messages as well

damianooldoni committed:
Add and set isScientificName field

damianooldoni committed:
Rename fields names useing camelCase and*InAnnex instead of *Original

damianooldoni committed:
set camelCase

damianooldoni committed:
Update changelog

damianooldoni committed:
Correct bug in if-else Jijnja template construction

niconoe commented:
Thanks @damianooldoni, I'll review this tomorrow!

niconoe commented on scripts/match_annexscientificname_to_scientificname.py:
```diff
@@ -0,0 +1,127 @@
+                                  {'scientificname_id': scientificname_id,
+                                   'id': row_id})
+
+def _remove_aux_columns(conn):
```
Maybe I'm missing something, but it seems a bit weird to create to columns in `create_new_tables` to finally remove them here. Are they temporary necessary in the meantime? Otherwise I'd suggest to not create them initially, and to remove this function.

damianooldoni replied:
I understand your point.

Yes, they are temporary necessary. 
Not using them, it means checking the presence of this names in `scientificname` on the flow while populating the table `annexscientificname` instead of populating the table first and then matching to `scientificname`.
I thought to split the two steps for making things cleaner but removing columns afterwards is not such clean.

damianooldoni replied:
I am refactoring by first adding the corrected version of the name (`scientificName`, `authorship`) from csv file to scientificname table, with the SQL INSERT-SELECT statement and then adding the other fields from file to anenxscientificname. As extra benefit we have a function less, `_update_scientificname_id()` as we don't have to update `scientificNameId` in db as we insert it in `annexscientificname` together with the other fields.

Pseudocode:

```
for each line in annex file:
    insert in scientificname and return id
    insert in annexscientificname
```

damianooldoni replied:
I have just realized why I opted to maintain the match to scientificname as a separate step: if they add some new annexes in database the match step can be performed autonomously. What do you think, @niconoe? I found it a good argument while implementing it.

damianooldoni replied:
Actually, I have to correct myself. As we removed the two columns in `match_annex_scientificname_to_scientificname()`, it was not reusable in maintenance phase as well. So, refactoring as we are doing is surely good. I will think to incapsulate code which will be useful for maintenance in a function so that it can be reusable later.

niconoe replied:
Yep, I guess having a single step at migration time (that relies on reusable functions that we can call from maintenance scripts) seems like the best approach for now. Do you agree?

niconoe commented on scripts/sql_snippets/create_new_tables.sql:
```diff
@@ -22,24 +22,26 @@ CREATE TABLE scientificname (
+    "matchType" gbifmatchtype,
+    CONSTRAINT scn_auth UNIQUE("scientificName", "authorship")
 );
+CREATE UNIQUE INDEX scn_auth_not_null ON scientificname("scientificName", ("authorship" IS NULL))
```
This works*, but I am wondering if we cannot simplify it to:

```
CREATE UNIQUE INDEX scn_auth_not_null ON scientificname("scientificName")
WHERE "authorship" IS NULL;
```

*: works: tested with simple INSERTS (reject second insert if author is null or not). I think the ON CONFLIC-returns None issue that you mentioned as the main problem is a different one - looking into it!

damianooldoni replied:
Thanks, I will test it. It makes sense.

damianooldoni committed:
Refactor by including match to scientificname table

See https://github.com/inbo/speciesbim/pull/49#pullrequestreview-508104095

damianooldoni committed:
Do not create two unneeded columns from annexscientificname

See https://github.com/inbo/speciesbim/pull/49#pullrequestreview-508104095

damianooldoni committed:
Simplify partial index

See https://github.com/inbo/speciesbim/pull/49#pullrequestreview-508132819

The version as proposed by @niconoe is easier to read and it seems to have same effect on data.

damianooldoni committed:
Move insert function to helpers as it can be used in maintenance

damianooldoni committed:
Add update_scientificname_id to helpers as it can bse useful for maintenance

damianooldoni commented:
Checked on ALL scientificnames and ALL names in annexes and it seems working well! :partying_face:

niconoe commented:
I can confirm that the whole transform_db process seems to work well with the new changes in both opened PR.

Will do a general code review now, but I think we're close to merge all this to the master branch.

Commit:
Minor refactoring

niconoe commented on scripts/populate_annex_scientificname.py:
```diff
@@ -11,25 +13,32 @@ def _get_annex(path):
+            is_scientific_name = True
+            if scientific_name_corrected == '':
+                is_scientific_name = False
+            annex_scientificnames[id] = {'id': id,
```
@damianooldoni: I notice you set the ID here in Python code, and it this ID is later pushed to the database like other fields... I think it would be a bit cleaner to not give any ID to the insert statement, so we let Postgres choose it (the field is set as a serial for this)

damianooldoni replied:
You are right. Do you change it or I?

niconoe replied:
I can do that!

niconoe replied:
Done, note that I also renamed `get_annex()` and simplified its implementation. 

In particular, it was returning a dict (key: ID, value: dict_of_other fields). It appeared the keys were actually not used later, only the values. So I simplified this data structure as a List of dicts.

niconoe commented on scripts/populate_annex_scientificname.py:
```diff
@@ -11,25 +13,32 @@ def _get_annex(path):
 
-def populate_scientificname_annex(conn, config_parser, annex_file):
-    """ Populate the table scientificnameannex
+def populate_annex_scientificname(conn, config_parser, annex_file):
```
(_this is a general comment, not to solved in this PR, but maybe useful for future developments or for a cleanup step_):

I notice some functions/loops like this one are quite long. I think it would be more readable to make them shorter, when possible. A few possible approaches in this case:

- Divide in smaller functions with a small and limited scope
- The code is sometimes longer and more difficult to read because the actual data logic is interleaved with logic to print/log what happens and code to measure execution time, I wonder if we can do something to minimize this. 

For the specific case of timing function calls, now that you're more and more comfortable with Python, you might be interested to have a look at [decorators](https://realpython.com/primer-on-python-decorators/). I think they can work well for this, see for example discussions and examples at https://stackoverflow.com/questions/51503672/decorator-for-timeit-timeit-method, https://stackoverflow.com/questions/34981602/logging-execution-time-with-decorators, https://medium.com/pythonhive/python-decorator-to-measure-the-execution-time-of-methods-fa04cb6bb36d, ...

Commit:
Refactoring (simplifying) annex loading

niconoe commented on scripts/populate_annex_scientificname.py:
```diff
@@ -0,0 +1,106 @@
+    logging.info(message_n_names_in_annex_file)
+
+    n_taxa_max = config_parser.get('annex_scientificname', 'taxa-limit')
+    if len(n_taxa_max) > 0:
+        n_taxa_max = int(n_taxa_max)
```
I'm a bit confused by this part, is it needed because of some data type issues (int vs str) in the configuration?

damianooldoni replied:
If `n_taxa_max` is empty in configuration, then import all names from annexes, otherwise take the first `n_taxa_max` names. And as everything from configuration is imported as characters, I make an integer of it.
